spec/unit/util/purgeable_spec.rb

```diff
@@ -30,7 +32,8 @@
 
     describe '#current_contents' do
       it 'collects direct contents of all managed directories' do
-        expect(subject.current_contents(recurse)).to contain_exactly(/\/expected_1/, /\/expected_2/, /\/unmanaged_1/, /\/unmanaged_2/, /\/managed_subdir_1/)
+        expect(subject.current_contents(recurse)).to contain_exactly(/\/expected_1/, /\/expected_2/, /\/unmanaged_1/, /\/unmanaged_2/,
+                                                                     /\/managed_subdir_1/, /\/managed_symlink_dir/, /\/unmanaged_symlink_file/)
```

Comment:
I'm a little confused about your test fixtures.. I assume some are symlinks to other ones, but it's not totally clear to me what's what. I'm also not sure why you've called some of them `unmanaged`, but then expect those files to be in this collection of managed directories.

Reply:
Note that this is a bug specifically for git-based environments, because Git does not follow symlinks when listing files it tracks: it tracks the symlink itself, but not what it points to, so the `desired_contents` that an environment passes to this class will contain the link but not the path to files that includes the link. So by calculating our contents differently than Git does, we introduced some overzealous purging by accident.

I don't know if other types of environments, which have a different way of collecting their desired contents, would behave the same way, but this is safe in those cases too. It's better to not purge something, and maybe leave around some cruft, than purge too much and risk nuking something a user needs (e.g. what if someone had a bad relative symlink that wound up pointing _outside_ the environment? Bad times).

Reply:
Yeah I added four symlinks, two pointing to dirs and two pointing to files, one managed and one unmanaged each. They're all relative, and they all point to fixture files that previously existed and are considered "desired". In general, these tests verify that "duplicate" paths to these desired files show up in the list of things to purge.

I'm working on another test right now that actually runs `purge!` on a copy of this fixture set, to prove that `rm_r` doesn't clean up anything it shouldn't when purging the unmanaged symlinks themselves.

Reply:
okay cool, that seems like a good test to add. otherwise I think this looks good!